### DISCUSSION TITLE
Larger than memory groupby and aggregate

### DISCUSSION THREAD

**Discussion by zhouzhou12** [duckdb/duckdb#6100] [Q&A]
I'm using the python api of duckdb, and encountering a difficulty.
How to group and aggregate data larger than memory? 
Joins can spill to hard disk when data can't fit into memory, can grouping aggregation?
Who can help me?
Thanks!

**Comment by Mytherin** [comment]
Aggregates work fine on larger than memory data if the hash table fits in memory. Spilling the hash table to disk is currently not yet supported but it is on our road map.

**Reply by zhouzhou12** [reply]
> Aggregates work fine on larger than memory data if the hash table fits in memory. Spilling the hash table to disk is currently not yet supported but it is on our road map.

Thanks for response so quickly! But I still have some doubts.

My data: 74202341 rows, 14 columns, 1.3GB, and column1 has 13173462 distinct values.

First, I do groupby column1 and aggregation.

sql:
CREATE TABLE IF NOT EXISTS temp1 AS
SELECT colomn1, 
    count(colomn2) item_num, 
    count(distinct colomn2) unique_item_num, 
    sum(colomn3) total_cnt, 
    sum(colomn4) total_amount
FROM t 
GROUP BY colomn1
order by colomn1;

memory usage: 24GB.

Then I limit the memory usage by using "PRAGMA memory_limit=20GB", it reports error:
"OutOfMemoryException: Out of Memory Error: could not allocate block of 262144 bytes (19999756288/20000000000 used) ".

As you said, Aggregates work fine on larger than memory data if the hash table fits in memory. It means that the hash table is larger than 20GB in this case?

**Reply by Mytherin** [reply]
The `count(distinct column2)` adds a requirement for more hash tables which increases the memory footprint. Perhaps try using `approx_count_distinct(column2)` instead?

**Reply by zhouzhou12** [reply]
> The `count(distinct column2)` adds a requirement for more hash tables which increases the memory footprint. Perhaps try using `approx_count_distinct(column2)` instead?

I tried using approx_count_distinct(column2), but the memory usage increased to over 100GB, and then crashed due to insufficient memory.

```sql
CREATE TABLE IF NOT EXISTS temp1 AS
SELECT colomn1,
    count(colomn2) item_num,
    approx_count_distinct(column2) unique_item_num,
    sum(colomn3) total_cnt,
    sum(colomn4) total_amount
FROM t
GROUP BY colomn1
ORDER BY colomn1;
```

**Reply by adriangb** [reply]
Is there a tracking issue I can follow? Larger than memory group by is what’s preventing me from using DuckDB in a specific workload.
